backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,202 @@
+package backtraceio.library.metrics;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.Message;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import backtraceio.library.common.BacktraceMathHelper;
+import backtraceio.library.common.BacktraceSerializeHelper;
+import backtraceio.library.interfaces.Api;
+import backtraceio.library.logger.BacktraceLogger;
+import backtraceio.library.models.BacktraceMetricsSettings;
+import backtraceio.library.services.BacktraceHandlerThread;
+import backtraceio.library.services.BacktraceReportSender;
+
+public abstract class BacktraceEventsHandler extends Handler {
+
+    private final static transient String LOG_TAG = BacktraceEventsHandler.class.getSimpleName();
+
+    protected final BacktraceHandlerThread backtraceHandlerThread;
+
+    /**
+     * Number of events in the queue
+     */
+    public int getCount() {
+        return events.size();
+    }
+
+    /**
+     * Maximum number of events in store. If number of events in store hit the limit
+     * BacktraceMetrics instance will send data to Backtrace.
+     */
+    private int maxNumEvents = 350;
+
+    public void setMaxNumEvents(int maxNumEvents) {
+        this.maxNumEvents = maxNumEvents;
+    }
+
+    public int getMaxNumEvents() {
+        return this.maxNumEvents;
+    }
+
+    /**
+     * Time between retries if metrics submission fails
+     */
+    private final int timeBetweenRetriesMillis;
+
+    /**
+     * List of events in the event queue
+     */
+    ConcurrentLinkedDeque<Event> events = new ConcurrentLinkedDeque<Event>();
+
+    /**
+     * Number of retry attempts
+     */
+    protected int numRetries = 0;
+
+    /**
+     * Http client
+     */
+    private final Api api;
+
+    /**
+     * Submission url
+     */
+    private final String submissionUrl;
+
+    /**
+     * User provided custom attributes
+     */
+    protected final Map<String, Object> customAttributes;
+
+    /**
+     * The application context
+     */
+    protected final Context context;
+
+    /**
+     * @param context
+     * @param api
+     * @param backtraceHandlerThread
+     * @param urlPrefix
+     * @param settings
+     */
+
+    public BacktraceEventsHandler(Context context, Map<String, Object> customAttributes,
+                                  Api api, final BacktraceHandlerThread backtraceHandlerThread, String urlPrefix, BacktraceMetricsSettings settings)
+    {
+        // This should always have a nonnull looper because BacktraceHandlerThread starts in the
+        // constructor and getLooper blocks until the looper is ready if the thread is started
+        super(backtraceHandlerThread.getLooper());
+        if (backtraceHandlerThread.getLooper() == null) {
+            throw new NullPointerException("Expected nonnull looper, this should not happen");
```

Comment:
wouldn't it be better if we will disable metrics instead? I'm thinking about the possible solutions without throwing an exception. It's fine to throw an exception when user makes a mistake (for example user passes a server URL that is null), but throwing it might disable the overall rest of our integration.

Reply:
@konraddysput sounds good, I think we can catch this exception from `EnableMetrics` method and just disable metrics if it doesn't work.

Reply:
@konraddysput we now catch exceptions in `EnableMetrics` method and disable metrics if we get an exception